Step1 - 인수 테스트 기반 리팩터링 리뷰 요청 드립니다.

programmer-sjk opened the pull request:
안녕하세요 😄 
이번에 `인수테스트 기반 TDD` 리뷰를 요청하게 된 서정국입니다.

`이번엔 꼭 TDD flow로 끝까지 해봐야지!!` 했지만 쉽지 않네요 😢 
어느새 도메인 로직을 수정하려는 제 자신을 발견하고는, 계속 도메인 테스트 코드를 개발하고 로직을 작성하도록 채찍질을 했습니다.

처음 리팩토링 할 때 욕심내서 수정을 많이 하다가 변경한 작업에 따른 영향이 커져 작업사항 날린 뒤, 작은 단위로 쪼개서 리팩토링을 진행했습니다. 확실히 조금씩 수정하고 리팩토링 하니 과제를 진행할 수 있었고, 인수테스트가 전체 기능을 검증해주니까 그래도 안심하고 리팩토링을 진행할 수 있었습니다.

리뷰에 도움이 되도록 작업 내용을 말씀드리면 아래와 같습니다.
* `순환참조` 가능성을 피하기 위해 repository 사용 (간단한 조회만 수행하기 때문에)
* 생성자 인자가 많은 경우 `빌더패턴` 사용
* Global Exception Handler 적용
* 에러 메시지 Enum 으로 관리 
* Sections `일급 컬렉션`, Distance 값 객체 활용
* Service에 transactional(readonly=true) 설정하여 오버라이딩
* 제너릭 타입 지정 
* `3단계 과제 (ATDD) 요구사항`과 달리, `기존 코드(5단계 인수테스트 기반 TDD)`에 지하철 역 삭제 시, **지하철 역이 없는 경우 예외**를 발생시키진 않더라구요. 그래서 뺄까? 하다가 삭제할 역이 없으면 예외처리 하는게 맞지 않을까? 싶어 그대로 반영했습니다. 

그럼 리뷰 부탁 드리겠습니다. 🙏

programmer-sjk committed:
docs(guide): README.md 요구사항 및 기능 구현 목록 작성

programmer-sjk committed:
test(acceptance): LineSectionAcceptanceTest 리팩터링

LineSectionAcceptanceTest 리팩터링
- LineSectionAcceptance 파일에 CRUD 함수 분리
- 통합 테스트 추가

programmer-sjk committed:
test(domain): Line의 getStations 함수에 대한 테스트 코드 작성

programmer-sjk committed:
feat(domain): Line의 getStations 함수 구현

programmer-sjk committed:
refactor(dto): LineRequest toLine 함수 리팩토링

programmer-sjk committed:
refactor(domain): Sections 1급 컬렉션 적용

programmer-sjk committed:
refactor(service): saveLine 메소드 리팩토링

programmer-sjk committed:
test(domain): Sections의 getStations 함수 테스트 코드 추가

programmer-sjk committed:
feat(domain): Sections의 getStations 기능 추가

programmer-sjk committed:
refactor(dto): LineResponse 리팩토링

programmer-sjk committed:
refactor(service, test): 순환참조를 피하기 위해 StationRepository 사용

programmer-sjk committed:
refactor(service): findLineById 함수 리팩토링

리팩토링
- private 접근 제한자 수정
- 조회 결과가 없을 때 NoResult 예외가 발생하도록 수정

programmer-sjk committed:
test(domain): 노선의 업데이트 함수 테스트 추가

programmer-sjk committed:
refactor(domain, test): 노선의 업데이트 함수 수정 및 관련 테스트 코드 수정

programmer-sjk committed:
refactor(service): @Transactional(readOnly = true) 적용

programmer-sjk committed:
test(domain): 노선 추가 기능의 domain 테스트 코드 작성

programmer-sjk committed:
feat(domain): 노선 추가 기능 관련 domain 기능 추가

programmer-sjk committed:
test(domain): 노선 삭제 기능 domain 테스트 추가

programmer-sjk committed:
refactor(domain, service): 노선 삭제 기능 리팩토링

programmer-sjk committed:
test(domain): Distance에 대한 테스트 작성

programmer-sjk committed:
refactor: Distance 객체를 프로덕션과 테스트 코드에 반영

programmer-sjk committed:
refactor: 코드의 에러메시지 Enum으로 관리하도록 수정

programmer-sjk committed:
refactor(controller): Generic 타입 명시

programmer-sjk committed:
refactor: GlobalException 핸들러 추가 및 관련 Http 상태 체크하는 테스트 코드 수정

programmer-sjk committed:
refactor: 사용하지 않는 import 문 제거

programmer-sjk committed:
refactor: LineRequest Builder 패턴 적용

programmer-sjk committed:
refactor(controller): 변수명 수정

programmer-sjk committed:
docs(guide): README.md 파일 수정

dhmin5693 commented on src/main/java/nextstep/subway/line/dto/LineResponse.java:
```diff
@@ -14,22 +16,38 @@ public class LineResponse {
     }
 
-    public LineResponse(Long id, String name, String color, List<StationResponse> stations, LocalDateTime createdDate, LocalDateTime modifiedDate) {
+    private LineResponse() {
```
불필요한 생성자를 사용하지 못하도록 막아두는 처리, 매우 좋습니다! 👍

dhmin5693 commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -2,181 +2,82 @@
-
-        return stations;
+    private Line findLineById(Long id) {
+        return lineRepository.findById(id).orElseThrow(NoSuchElementException::new);
```
`NoSuchElementException` 대신 적절한 예외를 사용하거나,
아니면 오류 메시지를 포함하는건 어떤가요?

운영 서버에서 문제가 발생할 때 굉장히 유용할거에요.

dhmin5693 commented on src/main/java/nextstep/subway/line/domain/Distance.java:
```diff
@@ -0,0 +1,39 @@
+import javax.persistence.Embeddable;
+
+@Embeddable
+public class Distance {
```
책임 구분 👍

dhmin5693 commented on src/main/java/nextstep/subway/line/domain/Distance.java:
```diff
@@ -0,0 +1,39 @@
+        return distance;
+    }
+
+    public Distance subtract(Distance newDistance) {
+        return new Distance(distance - newDistance.distance);
+    }
+
+    public Distance add(Distance another) {
+        return new Distance(distance + another.distance);
+    }
```
비슷한 역할을 수행하는데 매개변수명을 통일시켜도 괜찮지 않을까요?
개인적으론 another 가 조금 더 간결한 느낌은 있네요.

dhmin5693 commented on src/main/java/nextstep/subway/line/domain/Distance.java:
```diff
@@ -0,0 +1,39 @@
+
+    private void validate(int distance) {
+        if (distance <= ZERO) {
+            throw new IllegalArgumentException(ErrorCode.DISTANCE_BIGGEST_THAN_ZERO.getMessage());
```
오류 메시지를 정의하는 enum 클래스도 좋은 방법입니다.
코드의 응집성 면에서는 이 클래스가 상수로 오류 메시지를 가지고 있는게 괜찮지 않을까 싶은데 어떠신가요?

dhmin5693 commented on src/main/java/nextstep/subway/line/domain/Section.java:
```diff
@@ -50,23 +52,50 @@ public Station getDownStation() {
+        return Arrays.asList(upStation, downStation);
+    }
+
+    public void addLine(Line line) {
```
`Section` 의 입장에서 보면 `Line`을 추가하는게 아니라 소속이 변경된다고 볼 수 있습니다.
조금 더 적절한 메소드 명을 고려해보는건 어떨까요?
belong이나 included 같은 느낌은 어떠세요?

dhmin5693 commented on src/main/java/nextstep/subway/line/domain/Sections.java:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    private void validateHasBothStations(Section section) {
+        if (getStations().containsAll(section.getStations())) {
```
`List.containsAll(Collection)` 을 시간 복잡도의 측면에서 생각하면 어떨까요?
O(nm) 이 예상됩니다. (n: section에 포함된 station의 개수, m: stations에 포함된 station의 개수)

성능이 상당히 안좋지 않을까요?
받아온 getStations()을 `Set` 으로 변경한다면 O(n)으로 줄일 수 있을 것 같네요.

dhmin5693 commented on src/main/java/nextstep/subway/line/dto/SectionRequest.java:
```diff
@@ -1,14 +1,18 @@
     private Long upStationId;
     private Long downStationId;
-    private int distance;
+    private Distance distance;
```
`Distance` 는 도메인이지만 엔티티에 포함되는 요소이기도 합니다.
엔티티는 표현 영역에 드러내지 않는게 좋습니다.
만약 도메인이라는 목적, 즉 validation을 위함이라면 `Spring` 의 `@Valid` 를 활용해주세요.

또한 request는 코드로 작성되지 않은 외부로부터 데이터를 받는 부분인데요,
여기는 가급적 raw data만 사용하시는걸 추천드립니다.
(primitive type, String, 조금만 더 허용하면 enum까지 정도)

그 이유는 무엇이 들어올지 예측할 수 없는 범위에 있기 때문이에요.
raw data를 가져와서 내 코드가 알 수 있는 방법으로 직접 가공해주는게 예상치 못한 오류를 막는데 큰 도움이 됩니다.

dhmin5693 commented on src/test/java/nextstep/subway/line/domain/SectionTest.java:
```diff
@@ -0,0 +1,63 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class SectionTest {
```
테스트 커버리지를 보아하니 굉장히 촘촘하게 잘 작성해주셨어요. 좋습니다. 👍 

전 public으로 노출된 메소드(단순 getter 제외)는 모두 테스트를 작성해달라고 권장드리는 편입니다.
이 클래스에선 update, merge 중심으로 작성되어 있네요.

public은 왜 테스트가 필요할까요?
외부에서 사용하기 때문에 코드 사이 의존성이 생겨버리기 때문입니다.
의존성이 생긴 상황에서 public 메소드를 수정한다면요?
의존하는 곳에서 영향을 받지 않을 수가 없습니다.
그래서 어지간한 public은 전부 테스트를 작성해야 한다고 말씀드리는 것입니다.

내가 작성한 코드가 "정말 잘 작동한다"를 보장하기 위해서는 public 메소드까지,
그리고 public 메소드를 어떻게 사용해야 하는지 가이드하는 방식으로 작성하는게 좋습니다.

A의 상황에서 메소드를 수행하면 오류가 발생한다.
B의 상황에서 메소드를 수행하면 내부의 데이터가 ~~으로 변환된다.

이런 방식으로요.

dhmin5693 commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -2,181 +2,82 @@
 
 @Service
-@Transactional
+@Transactional(readOnly = true)
```
👍 

이것도 한번 읽어보세요~
https://www.baeldung.com/spring-transactions-read-only

programmer-sjk replied:
The Spring JPA repository marks all the methods as read-only transactions.
아 기본적으로 read-only 트랜잭션으로 처리하니 서비스 맨 위에 굳이 어노테이션을 붙일 필요가 없었군요!

dhmin5693 reviewed (changes requested):
정국님, 안녕하세요!
ATDD-TDD 미션을 함께하게 되었네요.
진심으로 반갑습니다. 👋👋👋
정국님께 더 좋은 경험으로 남을 수 있도록 최선을 다해볼게요!

---

고민한 부분을 꼼꼼히 정리해주셔서 리뷰에 큰 도움이 되었습니다.

커밋 내역을 보아하니 단위별로 나누기 위해 많은 노력을 기울이신 걸로 보여요.
테스트를 먼저 작성해야 하는 TDD는 그 방법 자체가 쉽지 않죠.
그러다보니 효용 여부 논쟁도 꽤 많기는 한 방법론입니다. 😁
그래도 미션이 미션이니만큼 시도하려고 하신 점을 높게 살 수 있겠네요 👏👏👏

코드 자체는 굉장히 잘 작성해주셨는데, 그래도 피드백을 남기지 않으면 아쉬우실 것 같아 reject 요청 한번 남겨봅니다.
확인해보시고 다시 요청주세요~

programmer-sjk committed:
refactor: 코드 리뷰 반영

코드 리뷰 반영
- NoSuchElementException 발생할 때 예외 메시지 추가
- containsAll -> set으로 수정
- domain public 테스트 코드 추가
- addLine -> belongs 로 메소드 이름 수정
- Request Dto에 Distance 멤버 변수 수정

programmer-sjk commented:
안녕하세요 동휘님!! 리뷰 주신 것 반영했습니다.
이전 인프라 미션때는 별 다른 리뷰없이 pass 했는데 이렇게 리뷰를 받게 되니 좋네요.

다른 것들은 다 반영했고 몇 가지 이야기 해 보고 싶은 것들을 남김니다.
**이야기 해 보고 싶은 것 1**
```
NoSuchElementException 대신 적절한 예외를 사용하거나,
아니면 오류 메시지를 포함하는건 어떤가요?
```
-> 우선 오류 메시지는 포함했습니다. 궁금한 게, `NoSuchElementException` 대신 어떤 에러가 적절하다고 보시나요? 사실 여기에 대해 조금 고민하고 검색을 해봤는데요. [스택오버플로우 글](https://stackoverflow.com/questions/24760314/is-there-a-standard-java-exception-class-that-means-the-object-was-not-found) 을 보고 `NoSuchElementException`와 `IllegalArgumentException` 두 예외 중에 고민하다가 `NoSuchElementException` 조금 더 데이터가 없음을 표현하는데 좋지 않을까? 라는 생각이 들어서 사용했는데 동휘님이라면 어떻게 하실지 궁금합니다. 

----

**이야기 해 보고 싶은 것 2**
```
오류 메시지를 정의하는 enum 클래스도 좋은 방법입니다.
코드의 응집성 면에서는 이 클래스가 상수로 오류 메시지를 가지고 있는게 괜찮지 않을까 싶은데 어떠신가요?
```
-> 사실 `미션 3단계 (ATDD)` 리뷰할 때만 해도 `에러 메시지를 Enum으로 관리하지 않았습니다`. 동휘님 의견처럼 **코드의 응집성이 도메인에 있는게 더 좋지 않을까** 생각했거든요. 헌데 3단계 리뷰어님이 이런 이야기를 남겨주셨습니다.
```
문자열로만 워싱해놓으면 어플리케이션에 에러핸들링, 메시지에 관련한 추가 개발이 있을때 
이를 직접 전부 찾아야하는 소요가 생기고 결국에는 변경이 불가피할 거 같아요!
```
이 의견을 듣고 생각해보니 에러 메시지가 수십, 수백개가 각 도메인에 있는데, **에러 메시지 앞에 공통적으로 `[ERRROR]` 문구를 추가**해야 하는 요구사항이 생긴다면? 이 의문을 가진 후 Enum으로 활용하는게 좋다는 생각을 했습니다. 그래서 수정하진 않았는데 동휘님 의견은 어떠신지 궁금합니다.

---

**이야기 해 보고 싶은 것 3**
```
전 public으로 노출된 메소드(단순 getter 제외)는 모두 테스트를 작성해달라고 권장드리는 편입니다.
이 클래스에선 update, merge 중심으로 작성되어 있네요.
```
-> 넵 저도 동의합니다. 이번에 domain 을 보고 놓친 테스트를 전반적으로 확인 후 추가했는데요.
**테스트 코드를 넣어야 할지 고민이 되는 코드**가 있어 이야기 해보고 싶었습니다. `Line 도메인`에 있는 delete 함수인데요
```
public void deleteSection(Station station) {
        sections.delete(station);
}
```
보시면 삭제를 `Sections`에 위임하고 있는데 테스트 코드를 작성해야 할까? 에 대해 3가지 생각이 있을 것 같아요
동휘님은 어떻게 생각하시나요?
* Sections 함수를 호출하니 `Sections의 delete 함수에 대해서만 테스트 코드`를 작성하면 되겠군
* 혹시 구현이 달라질지 몰라 `Line`의 `deleteSection` 함수에 대해서도 테스트 코드를 작성한다.
* Sections에 위임하니 테스트 코드는 작성하지 않지만, 만약 Line 도메인이 기능을 구현하게 된다면 그때 테스트 코드를 작성한다.

dhmin5693 commented:
### 1번
저는 커스텀 예외로 만드는걸 선호합니다.

### 2번
에러 메시지 앞에 [ERROR] 가 붙는다는 가정부터가 조금 애매하다고 생각되네요. API를 호출하는 쪽에서는 적절한 에러 코드(400 등)을 받고 오류로 인식하는게 조금 더 적절해 보이는데 굳이 에러 메시지까지 바꿀 필요는 없지 않을까요? 만약 그런 요구사항이 실제로 있다고 하더라도 지금 코드에 포함해주신 controller advice 내부에서 처리할 수 있어요. 매개변수에 `Exception` 타입을 넣어보세요.

저는 enum이 나쁜 방법은 아니라고 생각하지만 지금 말씀해주신 요구사항을 위해 enum을 사용할 필요는 없다고 봅니다.

### 3번
제가 제시드린 의견은 "내 코드가 정말 잘 작동함을 보장"함이었습니다. 그런 관점에서 본다면 `Line::deleteSection`의 동작이 `Sections`에 위임하고 있다 라는 것은 크게 중요하지 않습니다. 

자바를 공부할 때 가장 먼저 배우는 객체지향의 특징 중 **캡슐화와 정보 은닉**에 대해서 들어보셨을 겁니다. `Line` 외부에서 볼 때는 `Line`의 내부가 캡슐화되어 있어 정확히 알지 못하는게 정상입니다. 그저 `Line::deleteSection`를 호출하면 정거장이 삭제된다는 사실만을 알아야 합니다. 만약 호출부에서 getStations() 의 동작까지 알고 있다면, 메소드 호출 시 삭제했던 정거장이 반환된 정거장 목록에 포함되지 않으리라고 기대할 수 있겠죠. deleteSection 의 내부 동작은 언제든지 변할 수 있습니다. `Sections`에 의존하는건 그저 지금 구현되어 있는 방법일 뿐이라고 생각해주시는게 좋습니다. **캡슐화**되어 있기 때문에 내부의 구현 방법이 변해도 외부에선 알 수 없어야하며 그 동작은 변하지 않아야 하는 것이죠. 그런 논리라면 내부 구현이 어떻든간에 public으로 공개되어 외부에서 사용된 메소드는 테스트 코드로 동작을 보장하는게 맞습니다.

저는 정국님을 포함하여 다른 수강생 분들에게도 모든 public 메소드에 대해 테스트 작성을 했으면 좋겠다고 안내드리고 있습니다.
다만 작업량도 많고 교육이 다소 지루해질 수 있어 그렇게 **권장**하는 정도로만 말씀드리고 넘어가는 편입니다.

---

위의 답변과는 별개로, `Resolve conversation` 버튼을 너무 다 눌러주시진 않으셨으면 해요.
리뷰 시 완전히 처리된 내용은 resolve 처리를, 아직 처리가 덜 되었거나 추가 의견을 남기는 경우 버튼을 누르지 않는 방법으로 리뷰에 활용하고 있거든요.

dhmin5693 reviewed (approved):
피드백 잘 반영해주셨네요.
질문에 대한 답변 남겨두었으니 확인부탁드립니다.
2단계 미션 진행해주세요.
수고하셨습니다.
